bindings/src/main/java/io/pravega/storage/hdfs/HDFSChunkStorage.java

```diff
@@ -0,0 +1,360 @@
+            log.debug("Created '{}'.", fullPath);
+
+            // return handle
+            return ChunkHandle.writeHandle(chunkName);
+        } catch (FileAlreadyExistsException e) {
+            throw new ChunkAlreadyExistsException(chunkName, "HDFSChunkStorage::doCreate");
+        } catch (IOException e) {
+            throwException(chunkName, "doCreate", e);
+        }
+        return null;
+    }
+
+    @Override
+    protected boolean checkExists(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        ensureInitializedAndNotClosed();
+        FileStatus status = null;
+        try {
+            status = fileSystem.getFileStatus(getFilePath(chunkName));
+        } catch (IOException e) {
+            // HDFS could not find the file. Returning false.
+            log.warn("Got exception checking if file exists", e);
+        }
+        boolean exists = status != null;
+        return exists;
+    }
+
+    @Override
+    protected void doDelete(ChunkHandle handle) throws ChunkStorageException, IllegalArgumentException {
+        ensureInitializedAndNotClosed();
+        try {
+            boolean retValue = this.fileSystem.delete(getFilePath(handle.getChunkName()), true);
+        } catch (IOException e) {
+            throwException(handle.getChunkName(), "doDelete", e);
+        }
+    }
+
+    private void throwException(String chunkName, String message, IOException e) throws ChunkStorageException {
+        if (e instanceof FileNotFoundException) {
+            throw new ChunkNotFoundException(chunkName, message, e);
+        }
+        throw new ChunkStorageException(chunkName, message, e);
+    }
+
+    @Override
+    protected ChunkHandle doOpenRead(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        ensureInitializedAndNotClosed();
+        try {
+            FileStatus status = fileSystem.getFileStatus(getFilePath(chunkName));
+        } catch (IOException e) {
+            throwException(chunkName, "doOpenRead", e);
+        }
+        return ChunkHandle.readHandle(chunkName);
+    }
+
+    @Override
+    protected ChunkHandle doOpenWrite(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        ensureInitializedAndNotClosed();
+        try {
+            FileStatus status = fileSystem.getFileStatus(getFilePath(chunkName));
+            return ChunkHandle.writeHandle(chunkName);
+        } catch (IOException e) {
+            throwException(chunkName, "doOpenWrite", e);
+        }
+        return null;
+    }
+
+    @Override
+    protected int doRead(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset)
+            throws ChunkStorageException, NullPointerException, IndexOutOfBoundsException {
+        ensureInitializedAndNotClosed();
+        try {
+            if (fromOffset < 0 || bufferOffset < 0 || length < 0 || buffer.length < bufferOffset + length) {
```

Comment:
Same comment about preconditions and checking `fromOffset` and `bufferOffset`. make sure that we are comprehensively checking everything we need to check.

Reply:
Removed, this check was unnecessary. The preconditions are checked already.